feat: ssh keys db crud

kian99 opened the pull request:
## Description
This PR introduces add/list/remove methods to the `db` package for SSH keys. There is no read method because there is currently no use case where we need to lookup a single key.

Fixes [JUJU-7347](https://warthogs.atlassian.net/browse/JUJU-7347)

## Engineering checklist

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-7347]: https://warthogs.atlassian.net/browse/JUJU-7347?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
feat: add ssh keys database CRUD methods

alesstimec commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+// Copyright 2025 Canonical.
+package db
```
new line, please

alesstimec commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+	// It is expected that we only have 1 key that matches this fingerprint
+	// because of the unique constraint between users and public keys.
+	for _, key := range keys {
+		fp, err := calculateSSHFingerprint(key.PublicKey)
```
ah, yes, i think we should be storing the fingerprint.. then this method would be much simpler and fingerprint is a static piece of data that never changes for a given key

kian99 replied:
We can, I just initially took the route of not storing it since I figured users would have a small number of keys. I'll let others chime in then see about changing it.

kian99 replied:
Changed it to store the fingerprint.

alesstimec commented on internal/db/sshkeys_test.go:
```diff
@@ -0,0 +1,153 @@
+// Copyright 2025 Canonical.
+package db_test
```
new line, please

alesstimec commented on internal/db/sshkeys_test.go:
```diff
@@ -0,0 +1,153 @@
+		PublicKey:  []byte("foo"),
+		KeyComment: "bar",
+	}
+	err = s.Database.AddSSHKey(context.Background(), &key)
```
we should test what that we get a sensible error if we try adding the key twice

alesstimec reviewed (commented):
should we store key fingerprints?

kian99 committed:
chore: add test for duplicate key

SimoneDutto commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+		}
+	}
+
+	return errors.E(op, errors.CodeNotFound, "key not found")
```
```suggestion
	return errors.E(op, errors.CodeNotFound, return errors.E(op, errors.CodeNotFound, fmt.Sprintf("fingerprint for identity %s not found", identityName)))
```

SimoneDutto commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	if query.RowsAffected == 0 {
+		return errors.E(op, errors.CodeNotFound, "key not found")
```
```suggestion
		return errors.E(op, errors.CodeNotFound, fmt.Sprintf("comment for identity %s not found", identityName))
```

SimoneDutto commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+
+	var keys []dbmodel.SSHKey
+	if err := d.DB.Where("identity_name = ?", identityName).Find(&keys).Error; err != nil {
+		return errors.E(op, dbError(err))
```
i get this dbError but it would be better to add an identifier to these logs. So it's easier to troubleshoot in case of issue.

kian99 replied:
That's what the `op` does I believe

SimoneDutto replied:
i am referring to some identity identifier, like this `return errors.E(op, errors.CodeNotFound, fmt.Sprintf("comment for identity %s not found", identityName))`

kian99 replied:
I've added it to the remove key methods, but I don't think putting the user onto the error is necessarily needed for making debugging easier. Otherwise it just feels like we're adding all the input params onto the errors which can be excessive. Especially down in this method, maybe a method higher up can add more detail to the error in the future if needed.

SimoneDutto replied:
fair enough!

SimoneDutto commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+	return nil
+}
+
+// ListSSHKeys all a user's SSH keys.
```
```suggestion
// ListSSHKeys lists all user's SSH keys.
```

SimoneDutto commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// ListSSHKeys all a user's SSH keys.
+func (d *Database) ListSSHKeys(ctx context.Context, identityName string) (keys []dbmodel.SSHKey, err error) {
```
ListSSHKeysForUser i think would be better

kian99 replied:
A bit torn on this because the method signature has an `identityName` and none of the other methods are "forUser" or specify user.

SimoneDutto replied:
i think because the other ones are obvious, in the sense that you'd always a key for a user.
The listing can be done table wide, but here we are doing it per user

kian99 replied:
Fair enough, changing it.

kian99 committed:
chore: store public key fingerprint

Store the public key fingerprint in the DB rather than calculating it at runtime.

SimoneDutto reviewed (approved):
ty

kian99 committed:
Merge branch 'v3' into ssh-keys-db-crud

alesstimec commented on internal/db/sshkeys.go:
```diff
@@ -0,0 +1,98 @@
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	if err := d.DB.WithContext(ctx).Create(sshKey).Error; err != nil {
```
should we verify that the ssh key fingerprint is not empty?

alesstimec replied:
ah, i see we have a constraint..  i suppose it will be computed in the jimm package method (or the sshkey manager).

alesstimec commented on internal/db/sshkeys_test.go:
```diff
@@ -0,0 +1,160 @@
+	c.Assert(s.Database.DB.Model(&dbmodel.SSHKey{}).Count(&keyCount).Error, qt.IsNil)
+	c.Assert(keyCount, qt.Equals, int64(2))
+
+	err = s.Database.RemoveSSHKeyByFingerprint(context.Background(), "bob@canonical.com", gossh.FingerprintLegacyMD5(publicKey))
```
we should also test for "key not found" error here and in the next test to be aligned with juju's key handling responses.

alesstimec reviewed (approved):
LGTM with a few comments

kian99 committed:
Merge branch 'v3' into ssh-keys-db-crud
